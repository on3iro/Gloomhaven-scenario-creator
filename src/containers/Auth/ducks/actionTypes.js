export const LOGIN_SUBMIT = 'LOGIN_SUBMIT';
export const LOGIN_REQUEST = 'LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_ERROR = 'LOGIN_ERROR';
export const LOGOUT = 'LOGOUT';
