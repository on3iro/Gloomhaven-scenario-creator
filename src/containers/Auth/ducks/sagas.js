import axios from 'axios';
import { fork, call, put, takeLatest } from 'redux-saga/effects';

import * as types from './actionTypes';


// Login
export function *requestLogin(data) {
  const url = 'http://localhost:3030/api/auth/login';
  const config = {
    headers: {'Content-Type': 'application/json'},
  };

  try {
    const response = yield call(axios.post, url, data, config);

    yield put({ type: types.LOGIN_SUCCESS, payload: response.data });
  }catch (error) {
    yield put({ type: types.LOGIN_ERROR, payload: error.message });
  }
}

export function *authorize(action) {
  yield call(requestLogin, action.payload);
}

// Register
export function *requestRegister(data) {
  const url = 'http://localhost:3030/api/auth/register';
  const config = {
    headers: {'Content-Type': 'application/json'},
  };

  try {
    const response = yield call(axios.post, url, data, config);
    yield put({ type: types.REGISTER_SUCCESS, payload: response.data });
  }catch (error) {
    yield put({ type: types.REGISTER_ERROR, payload: error.response });
  }
}

export function *register(action) {
  yield call(requestRegister, action.payload);
}

// Auth watcher
export function *handleAuth() {
  yield takeLatest(types.LOGIN_SUBMIT, authorize);
  yield takeLatest(types.REGISTER_SUBMIT, register);
}

export default function *() {
  yield [
    fork(handleAuth),
  ];
}
