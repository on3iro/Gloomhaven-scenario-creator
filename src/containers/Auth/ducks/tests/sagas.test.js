import { takeLatest } from 'redux-saga/effects';
import moxios from 'moxios';

import * as sagas from '../sagas';
import * as types from '../actionTypes';


describe('handleAuthSaga', () => {
  const handleAuthSaga = sagas.handleAuth();

  it('should start task to watch LOGIN_SUBMIT', () => {
    const takeLatestDescriptor = handleAuthSaga.next().value;
    const expectedYield = takeLatest(types.LOGIN_SUBMIT, sagas.authorize);
    expect(takeLatestDescriptor).toEqual(expectedYield);
  });

  it('should start taks to watch REGISTER_SUBMIT', () => {
    const takeLatestDescriptor = handleAuthSaga.next().value;
    const expectedYield = takeLatest(types.REGISTER_SUBMIT, sagas.register);
    expect(takeLatestDescriptor).toEqual(expectedYield);
  });
});
