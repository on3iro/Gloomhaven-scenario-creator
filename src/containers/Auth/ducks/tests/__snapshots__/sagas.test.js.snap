exports[`handleAuthSaga should start taks to watch REGISTER_SUBMIT 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "REGISTER_SUBMIT",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`handleAuthSaga should start task to watch LOGIN_SUBMIT 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "LOGIN_SUBMIT",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`login authorize generator should call requestLogin with action.payload 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "test": "test",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`login authorize saga should call requestLogin with action.payload 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "test": "test",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`login requestLogin generator should call axios.post 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3030/api/auth/login",
      Object {
        "test": "test",
      },
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`login requestLogin generator should put LOGIN_ERROR action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": "Some error",
      "type": "LOGIN_ERROR",
    },
    "channel": null,
  },
}
`;

exports[`login requestLogin generator should put LOGIN_SUCCESS action with payload 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "email": "foo@bar.com",
        "id": 1,
        "name": "foo",
        "token": "abc123",
      },
      "type": "LOGIN_SUCCESS",
    },
    "channel": null,
  },
}
`;

exports[`login requestLogin should call axios.post 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3030/api/auth/login",
      Object {
        "test": "test",
      },
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`login requestLogin should put LOGIN_ERROR action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": "Some error",
      "type": "LOGIN_ERROR",
    },
    "channel": null,
  },
}
`;

exports[`login requestLogin should put LOGIN_SUCCESS action with payload 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "email": "foo@bar.com",
        "id": 1,
        "name": "foo",
        "token": "abc123",
      },
      "type": "LOGIN_SUCCESS",
    },
    "channel": null,
  },
}
`;
